src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+#include "ServicePool.h"
+
+#include <string.h>
+
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/RandUtils.h>
+
+using chip::Mdns::kMdnsTextMaxSize;
+using chip::Mdns::MdnsService;
+using chip::Mdns::TextEntry;
+
+namespace {
+
+void FreeTextEntries(MdnsService * service)
+{
+    if (service->mTextEntryies)
+    {
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            TextEntry & entry = service->mTextEntryies[i];
+
+            if (entry.mKey)
+            {
+                chip::Platform::MemoryFree(const_cast<char *>(entry.mKey));
+                entry.mKey = nullptr;
+            }
+            if (entry.mData)
+            {
+                chip::Platform::MemoryFree(const_cast<uint8_t *>(entry.mData));
+                entry.mData = nullptr;
+            }
+        }
+        chip::Platform::MemoryFree(service->mTextEntryies);
+        service->mTextEntryies = nullptr;
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntryies = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        dest->mTextEntrySize = source.mTextEntrySize;
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+            size_t keySize    = strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize);
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
`strnlen` returns `kMdnsTextMaxSize` when it doesn't find a \0 in the `kMdnsTextMaxSize` bytes of the string. This check is to eliminate such possibility since we don't want to make a truncated copy.

Reply:
Oh, I see, that's addressing the `strncpy` issue.  But is `kMdnsTextMaxSize` the max _buffer_ size or the max size of the text itself?  Other places compare that many chars, which suggests it's the text size, but here we are treating it effectively as the buffer size?  WE should probably be consistent about this, and document the constant's meaning.